### PR TITLE
Feature/DACT-1244 JS Validations for correspondence manual address page 

### PR DESCRIPTION
**Pull request by mdesaikns**
Also added config for residential address to make follow up changes moderately small.

### CODE REVIEW

File: src/controllers/director.correspondence.address.manual.controller.ts [hunk 1]
```diff
@@ -158,3 +159,22 @@ export const buildValidationErrors = (validationStatusResponse: ValidationStatus
...
+    templateName: Templates.DIRECTOR_CORRESPONDENCE_ADDRESS_MANUAL,
+    backLinkUrl: urlUtils.getUrlToPath(DIRECTOR_CORRESPONDENCE_ADDRESS_SEARCH_PATH, req),
+    directorName: formatTitleCase(retrieveDirectorNameFromFiling(officerFiling)),
+    typeahead_array: COUNTRY_LIST,
+    correspondence_address_premises: serviceAddress.premises,
+    correspondence_address_line_1: serviceAddress.addressLine1,
+    correspondence_address_line_2: serviceAddress.addressLine2,
+    correspondence_address_city: serviceAddress.locality,
+    correspondence_address_county: serviceAddress.region,
+    typeahead_value: serviceAddress.country,
```
Comment (okoye-companieshouse) [inline src/controllers/director.correspondence.address.manual.controller.ts:175]: I am confused here and don't know if it is a requirement but why is the country called typeahead_value?

File: src/controllers/director.correspondence.address.manual.controller.ts [hunk 2]
```diff
@@ -158,3 +159,22 @@ export const buildValidationErrors = (validationStatusResponse: ValidationStatus
...
+    correspondence_address_postcode: serviceAddress.postalCode,
+    typeahead_errors: JSON.stringify(formattedErrors),
```
Comment (okoye-companieshouse) [inline src/controllers/director.correspondence.address.manual.controller.ts:177]: Are these errors for the typeahead_value(country)?

File: src/validation/manual.address.validation.ts [hunk 3]
```diff
@@ -0,0 +1,220 @@
+import { Address } from "@companieshouse/api-sdk-node/dist/services/officer-filing";
+import { ManualAddressValidationType, ValidationError } from "../model/validation.model";
+import { COUNTRY_LIST, UK_COUNTRY_LIST } from "../utils/properties";
+
+const REGEX_FOR_MANUAL_ADDRESS = "^[-,.:; 0-9A-Z&@$£¥€'\"«»?!/\\\\()\\[\\]{}<>*=#%+ÀÁÂÃÄÅĀĂĄÆǼÇĆĈĊČÞĎÐÈÉÊËĒĔĖĘĚĜĞĠĢĤĦÌÍÎÏĨĪĬĮİĴĶĹĻĽĿŁÑŃŅŇŊÒÓÔÕÖØŌŎŐǾŒŔŖŘŚŜŞŠŢŤŦÙÚÛÜŨŪŬŮŰŲŴẀẂẄỲÝŶŸŹŻŽa-zſƒǺàáâãäåāăąæǽçćĉċčþďðèéêëēĕėęěĝģğġĥħìíîïĩīĭįĵķĺļľŀłñńņňŋòóôõöøōŏőǿœŕŗřśŝşšţťŧùúûüũūŭůűųŵẁẃẅỳýŷÿźżž]*$";
```
Comment (okoye-companieshouse) [inline src/validation/manual.address.validation.ts:5]: Should these regex's not be neater from another validation util file to make them reusable in the case of testing?

File: src/validation/manual.address.validation.ts [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+	// City
+	if(!address.locality) {
+		validationErrors.push(manualAddressValidationType.MissingValue.City);
+	} else {
+		validationErrors = validateCity(address.locality, manualAddressValidationType, validationErrors);
+	}
+
+	// County, Town, Province, Region
+	if(address.region) {
```
Comment (okoye-companieshouse) [inline src/validation/manual.address.validation.ts:41]: Were you meant to check for the presence or absence of region?

File: src/controllers/director.correspondence.address.manual.controller.ts (hunk 1 above)
Comment (mdesaikns) [reply src/controllers/director.correspondence.address.manual.controller.ts:175]: Hi @okoye-companieshouse , I have _not_ changed the functionality of this block of code, if you notice I've moved block implementation in `renderPage()`method (copied code from 82-97).

File: src/controllers/director.correspondence.address.manual.controller.ts (hunk 2 above)
Comment (mdesaikns) [reply src/controllers/director.correspondence.address.manual.controller.ts:177]: no, these are for all the fields on the page, similar to the other comment, the implementation has not changed, it moved to a separate method.

File: src/model/validation.model.ts [hunk 5]
```diff
@@ -45,6 +45,36 @@ export interface OccupationValidationType {
     }
 }
 
+export interface ManualAddressValidationType {
```
[the remark this answers is not in the document]
Comment (mdesaikns) [reply src/model/validation.model.ts:48]: I will explore this, however, I'd prefer to have this as it's similar to other validation configs which provides greater readability across services.

File: src/validation/manual.address.validation.ts (hunk 4 above)
Comment (mdesaikns) [reply src/validation/manual.address.validation.ts:41]: no, it's an optional field.

File: src/validation/manual.address.validation.ts (hunk 3 above)
Comment (mdesaikns) [reply src/validation/manual.address.validation.ts:5]: Yes, I agree to this. There have been some inconsistencies across some ACs covering character validation. I will compare this with occupation and move it to util class.

**Review by okoye-companieshouse — APPROVED** [review]

File: src/validation/manual.address.validation.ts [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+export const validateCity = (city: string, manualAddressValidationType: ManualAddressValidationType, validationErrors: ValidationError[]): ValidationError[] => {
+	if(!city.match(REGEX_FOR_MANUAL_ADDRESS)) {
+		validationErrors.push(manualAddressValidationType.InvalidCharacters.City);
+		return validationErrors;
+	}
+
+	if(city.length > ALLOWED_LENGTH) {
+		validationErrors.push(manualAddressValidationType.InvalidLength.City);
+		return validationErrors;
```
[the remark this answers is not in the document]
Comment (mdesaikns) [reply src/validation/manual.address.validation.ts:127]: hi @okoye-companieshouse , yes we had to follow an order of priority for validation errors, which forces us to return a validation error if a higher-order validation error is encountered.

**Pull request merged**